[addons][inputstream][videocodec] rework documentations, fix "C" ABI and fix memleaks

AlwinEsch opened the pull request:
## Description
There are some changes that fix Memleaks that I noticed on the addons.

This is ensured by the fact that the "C" structures at the input stream are not transferred to the addon but are managed using C++ classes, which are then used by the addon.

In addition, some things are simplified because the data does not have to be used on the addon using "C".

The other, and most importantly with Memleaks, is that the interface to inputstream and videocodec is now 100% "C" and no "C++" stuff is given between Kodi and addon.

I will go through this change step by step on all addon types, on the one hand to make feature requests in this regard (where users also requested a usable "C") and to be really sure that there are no creeping errors in the "C" ABI.

In the end (other requests), a "C" ABI test system is brought to Kodi to ensure that nothing wrong gets into it.

Regarding documentation, you can see here how it looks:
- http://alwinesch.github.io/group__cpp__kodi__addon__inputstream.html
- http://alwinesch.github.io/group__cpp__kodi__addon__videocodec.html

There are certainly a lot of grammatical errors and also certainly wrong descriptions, if someone could look over it and my miserable English would be correct :grinning:.

Also added on first commit the support about BlockSize where on stream reads is required.

**This is still WIP and certain things are not yet correct.**

As a note: I have adapted the changes to our code guidelines.

**EDIT** 
I forgot a part of the codec where the "C" structures are in their own memory (to simplify API changes, like on https://github.com/xbmc/xbmc/pull/17438/commits/c16238230e66adb106c9b460886e6939f4175dd2)

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented:
Is now cleaned up by add of code from other requests, further is on #17501 another separated part created.

After them is in I look to take some documentation parts where become no change from API here.

AlwinEsch committed:
[addons][inputstream] change get_stream_ids call to match "C" API

Before it contains a C++ part where not match API.

Also id the C++ addon call changed to use std::vector

AlwinEsch committed:
[addons][inputstream] add special callback on GetStream function

This is added to prevent memleaks and to use stack memory.

This becomes on next changes improved and done by a C++ structure
inside Headers.

AlwinEsch committed:
[addons][inputstream] add C++ InputstreamCapabilities to translate INPUTSTREAM_CAPABILITIES

This is added to translate the "C" structure INPUTSTREAM_CAPABILITIES to C++
class InputstreamCapabilities.

This added to make access easier and more secure against leaks.

AlwinEsch committed:
[addons][inputstream] use C++ class for INPUTSTREAM_MASTERING_METADATA and INPUTSTREAM_CONTENTLIGHT_METADATA

This to have them more clean and to match more the interface.

AlwinEsch committed:
[addons][inputstream] add C++ InputstreamTimes to translate INPUTSTREAM_TIMES

This is added to translate the "C" structure INPUTSTREAM_TIMES to C++
class InputstreamTimes.

This added to make access easier and more secure against leaks.

AlwinEsch commented:
jenkins build this please

Have cleaned up the request and mostly done from my side, now only cleanups and improvements inside.

Is hard to make them as good as possible as the stream is also a major part of Kodi and a major part of related addons.

The biggest reasons why I see no way around are:
- About Kodi Trademark problems with binary addons if another place want change his Kodi name
  - Is hard for binaries to use different names (about include, hardcodec parts and cmake packages) and not use much hacks for work
  - Eventually the addon folder then indpendent from Kodi
  - This a part where I start soon with github project here to find a way together with Team
- To confirm always a safe "C" ABI between addon and Kodi
  - Before was always come from time to time a "C++" part inside where breaks it
  - Also should in future the addon API usable as pure "C" for another languages
    - This currently still not possible, as still some "C++" parts in "C" places

Bad is, as also the DEMUX_PACKET is PVR related need every addon there a version update (and where used code change) :unamused: :smirk: Informed @ksooo some time ago about.

Ping @peak3d @phunkyfish

AlwinEsch commented:
I look to split this request in smaller ones, only bad that always related addons need Version increase.
But think better as here to much for hold overview.

AlwinEsch commented:
Due to speak with @peak3d and about GSOC this stays open and maybe comes in 4 weeks.

AlwinEsch committed:
[addons][inputstream] move StreamCodec.h to inputstream folder

Further is his "C" code on "kodi/c-api/addon-instance/inputstream/stream_codec.h"
now. This to have them more clean as it only relate to inputstream and codec
addon instances and not to other addon types.

AlwinEsch committed:
[addons][inputstream] make StreamCrypto.h in own folder and improve

There is the "C" part about moved to "kodi/c-api/addon-instance/inputstream/stream_crypto.h"
and parts renamed to have more clean. Further is the C++ helper class
"kodi::addon::StreamCryptoSession" added to use C++ on addons about.

AlwinEsch committed:
[addons][pvr] use new "C" style DEMUX_PACKET for streams

Before was for streams always Kodi's C++ based structure
used where not match a "C" ABI for the addons.

AlwinEsch committed:
[addons][player] move down the previous on addons used headers

This are no more used now by addons itself and only be used on
videoplayer itself.

AlwinEsch committed:
[addons][codec] add C++ kodi::addon::VideoCodecInitdata

Class to handle the "C" structure VIDEOCODEC_INITDATA in C++.

AlwinEsch committed:
[addons][inputstream][codec][pvr] increase API versions

This increase all changed addon instance versions.
PVR included as it use a the new DEMUX_PACKET structure now.

AlwinEsch committed:
[addons][inputstream] change interface headers to match and support "C" style

Before was a pure "C" compile not possible, which shows that this ABI is not set correctly.

The interface must be sure that it is only in pure "C"!

This change only in relation to inputstream, all other types and positions also require
some changes.

garbear commented on xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp:
```diff
@@ -367,8 +367,10 @@ std::vector<CDemuxStream*> CInputStreamAddon::GetStreams() const
 {
-  INPUTSTREAM_INFO stream = m_struct.toAddon->get_stream(&m_struct, streamId);
-  if (stream.m_streamType == INPUTSTREAM_INFO::TYPE_NONE)
+  INPUTSTREAM_INFO stream;
+  memset(&stream, 0, sizeof(INPUTSTREAM_INFO));
```
Why not just

```c++
INPUTSTREAM_INFO stream{};
```
this should have the same effect, right?

AlwinEsch replied:
You're right, it's a holdover from the old days and no longer necessary.

Still partially overlook this one and will change it.

garbear commented on xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/Inputstream.h:
```diff
@@ -14,11 +14,49 @@
+  const std::map<std::string, std::string> GetProperties() const
+  {
+    std::map<std::string, std::string> props;
+    for (unsigned int i = 0; i < m_cStructure->m_nCountInfoValues; ++i)
+      props.insert(std::make_pair<std::string, std::string>(
+          m_cStructure->m_ListItemProperties[i].m_strKey,
+          m_cStructure->m_ListItemProperties[i].m_strValue));
```
Minor comment, I would say that the for loop is a single command, but multiple lines, and would qualify for `{}`. I think the additional vertical spacing would also aid readability.

garbear commented on xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/inputstream/TimingConstants.h:
```diff
@@ -0,0 +1,34 @@
+  return x * STREAM_TIME_BASE;
+}
+
+constexpr double STREAM_MSEC_TO_TIME(double x)
```
Nice use of `constexpr`

garbear commented on xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/inputstream/stream_constants.h:
```diff
@@ -0,0 +1,50 @@
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
```
I don't believe `#pragma once` is appropriate C. Header guards are the idiomatic way to prevent double inclusion of C headers

garbear replied:
It's good if I'm wrong, because I much prefer `#pragma once`

AlwinEsch replied:
There I forget on most headers to add, e.g.:
```c
#ifndef C_API_ADDONINSTANCE_INPUTSTREAM_STREAMCONSTANTS_H
#define C_API_ADDONINSTANCE_INPUTSTREAM_STREAMCONSTANTS_H

#endif /* !C_API_ADDONINSTANCE_INPUTSTREAM_STREAMCONSTANTS_H */
```

garbear commented on xbmc/cores/VideoPlayer/Interface/Addon/DemuxPacket.h:
```diff
@@ -9,42 +9,38 @@
+      recoveryPoint = false;
+
+      cryptoInfo = nullptr;
+    }
```
I see what you did here with the struct inheritance. That's pretty cool.

garbear reviewed (approved):
Good work. Are you excited the the API is fully coming together?

AlwinEsch commented:
Also seen on review, still some things todo on this request :sweat_smile:, was for now only a rebase.

But after here it makes no more matter with what addon was created :sunglasses:

AlwinEsch committed:
[addons][codec] separate "C++" and "C" parts in own headers

This to cleanup and to confirm "C" safe ABI and base for other languages.
Also is on the C++ header a clang cleanup done.

AlwinEsch committed:
[addons][inputstream][codec] rework addon documentation

This makes the documentation about inputstream and videocodec to style
needed for binary addons and to show correct on Website.

There still some parts where need improved and a bit more detailed
text, but think to make indpendent then as it not touch API itself.

AlwinEsch committed:
[addons][inputstream] change get_stream call to match C API

Before was there the complete structure has return value where makes
harder in unused cases and not match a correct "C" ABI.

AlwinEsch committed:
[addons][inputstream] separate "C" and "C++" header parts in own files

This is done to confirm a pure "C" within interface and to allow
use for other languages where need a "C" as base.

This changes are the initial part, there still few included headers
where use C++!

This becomes improved on following changes, for here to have the
other commits more clean and see if relate to "C" between Kodi
or to "C++" on addon.

AlwinEsch committed:
[addons][inputstream] add C++ InputstreamInfo to translate INPUTSTREAM_INFO

This is added to translate the "C" structure INPUTSTREAM_INFO to C++
class InputstreamInfo.

This added to make access easier and more secure against leaks.

AlwinEsch committed:
[addons][inputstream][videoplayer] separate stream structures and parts to addon dev-kit

There becomes the parts where was previous on Kodi's videoplayer moved to
the addon dev kit and much cleaned to confirm the "C" ABI style.

This change is not nice but needed for this:
- The dev-kit should be soon independent of Kodi (trademark, other languages, smaller)
- Before was there some C++ parts inside where breaks a safe ABI use between addon and Kodi
- Have the header places for addons better sorted and more clean in view

AlwinEsch committed:
[addons][inputstream] include InputStreamConstants.h inside addon headers dir

This add the header to the kodi-dev-kit folder to have them all
together and prepared for a independent dev kit, further to confirm
more the "C" use between Kodi and addon.

AlwinEsch committed:
[addons][inputstream] use independent TimingConstants.h for addons

This separate the Kodi TimingConstants.h to use a own on addons
where his text begin with "STREAM_..." now and them with "DVD_..." stays
on Kodi itself.

There are few Pro's and Contra's about but for wanted API better to have.

+++: It stays in pure "C" for them and as base for other languages and the "C++" also on new header
+++: For the time where the headers comes into a own dev-kit and no more in Kodi source
+++: No file outside of the addon header placed in Kodi
---: On changed by them in Kodi must be them on addon headers also by value changes the header updated.

phunkyfish commented:
jenkins build this please

phunkyfish commented:
jenkins build this please

AlwinEsch committed:
[addons][inputstream] add C++ InputstreamProperty to translate INPUTSTREAM_PROPERTY

This is added to translate the "C" structure INPUTSTREAM_PROPERTY to C++
class InputstreamProperty.

This added to make access easier and more secure against leaks.

Rechi commented on stream_constants.h.diff:
```diff
@@ -0,0 +1,18 @@
+--- a/xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/inputstream/stream_constants.h
```
What's this file used for?

AlwinEsch replied:
Seems to come in on rebase, Sorry!

Becomes today kicked out and think to add also on .gitignore a `/*.diff` below of `/*.patch` to prevent it on base folder in future.
